include/pybind11/pytypes.h

```diff
@@ -377,20 +380,49 @@ class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
     /// Constructs a new exception from the current Python error indicator, if any.  The current
     /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    error_already_set() : std::runtime_error("") {
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        if (m_type) {
+            PyErr_NormalizeException(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
```

Comment:
Was there a discussion about this already?
It would seem better to also do this lazily.
https://docs.python.org/3/c-api/exceptions.html#c.PyErr_NormalizeException
Quote: The delayed normalization is implemented to improve performance.

The documentation isn't clear about it, but I know from recent experience that this function can fail. I believe we need to check `PyErr_Occurred()` with `pybind11_fail`, or even `std::terminate` (similar to how an exception during unwinding causes process termination). Again from recent experience, original errors getting masked by errors in the error handling can be extremely time consuming to debug. Better kill the process, reporting as much information as possible.

Reply:
Hm ... OK, let's keep this for later.

Based on the very-real-to-me "pain" experience I mentioned earlier, I'm actually pretty worried about the case that `PyErr_NormalizeException()` fails, but apparently that's not handled here, or in type_caster_base.h. But then again, that's an existing issue that we (me) can look into separately.